Support domain-only EthereumSignTypedData and other improvements

aloisklink opened the pull request:
Add documentation on using `EthereumSignTypedData` on T1 and improved types. These were missing from #1015.

### Minor breaking change

`EthereumSignTypedData` now requires passing Trezor T variables. This seems makes sense from my side, since everybody using this function should pass Trezor T variables anyway, and since this function was only recently released, I doubt this change would affect any users of this library.

Plus, it makes all the code (especially types!) simpler.

Trezor Model 1 parameters are still optional, as blind signing requires using an extra dependency on an external library, and is less secure.

### Support for `primaryType` = `EIP712Domain`

~I tried adding docs/tests/types for https://github.com/trezor/trezor-firmware/pull/2036 in this PR too, but updating the `submodules/trezor-common` submodule just causes a bunch of other Protobuf type errors to occur, so I'll do that in another separate PR in the future.~

I've added support for https://github.com/trezor/trezor-firmware/pull/2036 in this PR too, since they both edit similar files.

I've tested locally using:

- `./tests/run.sh -D 'podman' -i 'ethereumSignTypedData'` Success now #1042 been merged ~Fails~
- `./tests/run.sh -D 'podman' -i 'ethereumSignTypedData' -f '2-master'` Success
- `./tests/run.sh -D 'podman' -i 'ethereumSignTypedData' -f '1-master'` Success

aloisklink committed:
refactor(ethSignTypedData)!: Improve docs & types

BREAKING CHANGE: EthereumSignTypedData now requires passing Trezor T
  variables. Trezor Model 1 parameters are still optional, as
  blind signing requires using an extra dependency on an external
  library, and is less secure.

aloisklink commented on src/js/core/methods/EthereumSignTypedData.js:
```diff
@@ -9,15 +9,17 @@ import type { MessageResponse, EthereumTypedDataStructAck } from '../../types/tr
+        address_n: number[],
+        ...$Exact<EthereumSignTypedDataParams> | $Exact<EthereumSignTypedHashOrDataParams>,
+    },
+    { path: any }, // removes the "path" variable from this.params
```
The `path` value in `this.params` is never used, and just causes Flowtype issues, so I removed it using [`$Diff<>`](https://flow.org/en/docs/types/utilities/#toc-diff)

aloisklink commented on src/js/core/methods/EthereumSignTypedData.js:
```diff
@@ -45,13 +47,18 @@ export default class EthereumSignTypedData extends AbstractMethod<'ethereumSignT
+            data: payload.data,
         };
+
+        if (payload.message_hash) {
+            this.params = {
+                ...this.params,
+                domain_separator_hash: payload.domain_separator_hash,
+                message_hash: payload.message_hash,
+            };
+        }
```
This is in a separate `if`-statement to make Flow-type happy.

aloisklink commented on src/js/core/methods/EthereumSignTypedData.js:
```diff
@@ -84,9 +93,7 @@ export default class EthereumSignTypedData extends AbstractMethod<'ethereumSignT
             };
         }
 
-        validateParams(this.params, [{ name: 'data', type: 'object', required: true }]);
```
No longer needed in  the `run()` command, since these Trezor T parameters are now always required, and so this is validated in the `init()` function.

mroz22 commented:
Hi, we are switching to master branch of trezor-common here https://github.com/trezor/connect/pull/1036/files and yes. It needs to sort out some types incompabilities. Patience please.

mroz22 commented:
Updated trezor-common is now in develop branch.

aloisklink committed:
fix(ethSignTypedData): Support domain-only signs

Add support for signing domain-only data in ethereumSignTypedData,
e.g. when primaryType = EIP712Domain.

On Trezor T, this behaviour previously gave incorrect signatures
or threw errors.

On Trezor Model 1, not passing the "message_hash" parameter caused
an error.

mroz22 commented:
Hello @aloisklink 
We are generally quite happy for your contributions as they are always really nicely done, so If it suits your needs better this wa  I think I am not going to object against that. Just if you could please rebase this on the latest develop so that I can run T1 tests in gitlab CI (we have added better support for this in one of the last commits).

mroz22 commented on src/js/types/networks/ethereum.js:
```diff
@@ -101,22 +101,34 @@ type EthereumSignTypedDataMessage<T: EthereumSignTypedDataTypes> = {
+    ...$Exact<EthereumSignTypedData>,
+    ...$Exact<EthereumSignTypedHash>,
+};
+
```
can we maybe have this also in ethereum.d.ts?

aloisklink replied:
Done in https://github.com/trezor/connect/commit/7139d792a495d4db5cfd20ae63d1e1c3c03539a9

I've also renamed this type to `EthereumSignTypedHashAndData` instead of `...HashOrData`, to make it more clear that it should have both the typed hash AND the typed data.

aloisklink commented:
> We are generally quite happy for your contributions as they are always really nicely done, so If it suits your needs better this wa I think I am not going to object against that. Just if you could please rebase this on the latest develop so that I can run T1 tests in gitlab CI (we have added better support for this in one of the last commits).

No worries, this isn't urgent, this PR is currently mainly just documentation/improved-types.

I'm also very close to being done with adding support for https://github.com/trezor/trezor-firmware/pull/2036 (I can push it into this PR, or into a separate one), which fixes any of the failing tests in `ethereumSignTypedData` due to updating `trezor-common`.

I did encounter a bug with `getFirmwareRange()`: it only uses the first matching entry in `src/data/config.json`, but I believe I've fixed it. I'm just testing it locally, I'll make a PR in a few hours!

aloisklink commented:
> Just if you could please rebase this on the latest develop so that I can run T1 tests in gitlab CI (we have added better support for this in one of the last commits).

Rebased on the latest `develop` in fee63892ad2ac1f85156af302249f2f5a3ac04a3

I did it last night before you commented, so GitHub's UI is showing the commit before your comment.

I also added support for https://github.com/trezor/trezor-firmware/pull/2036 in https://github.com/trezor/connect/pull/1033/commits/0bff3e530364ae5b8d9cae70a385a9680450aef8, so that the T1/TT tests now pass on the updated fixtures. **Let me know if you'd prefer me to put this into a separate PR.**

I've tested locally using:

- `./tests/run.sh -D 'podman' -i 'ethereumSignTypedData'` Fails, since `getFirmwareRange()` isn't throwing a wrong firmware error (see #1042)
- `./tests/run.sh -D 'podman' -i 'ethereumSignTypedData' -f '2-master'` Success
- `./tests/run.sh -D 'podman' -i 'ethereumSignTypedData' -f '1-master'` Success

## Fixup commits

- https://github.com/trezor/connect/pull/1033/commits/2fbfc671d1147720283d0eb57e9ffa260457edc8 Fixes the unit test for `getUnavailableCapabilities()` (I didn't realise that changing `src/data/config.json` would affect this test)
- https://github.com/trezor/connect/pull/1033/commits/7139d792a495d4db5cfd20ae63d1e1c3c03539a9 Renames flow-type `EthereumSignTypedHashOrData` to `EthereumSignTypedHashAndData` and adds it to the TypeScript definitions too.

aloisklink committed:
fix(plugins/ethereum): Support domain-only EIP712

Adds support for domain-only EIP-712 hashes to
plugins/ethereum/typedData.js

aloisklink commented on src/js/plugins/ethereum/typedData.js:
```diff
@@ -20,6 +20,17 @@ function sanitizeData(data) {
     }
 }
 
+/**
+ * Calculates the domain_separator_hash and message_hash from an EIP-712 Typed Data object.
+ *
+ * The Trezor Model 1 does not currently support constructing the hash on the device,
+ * so this function pre-computes them.
+ *
+ * @template {sigUtil.TypedMessage} T
+ * @param {T} data - The EIP-712 Typed Data object.
+ * @param {boolean} metamask_v4_compat - Set to `true` for compatibility with Metamask's signTypedData_v4 function.
+ * @returns {{domain_separator_hash: string, message_hash?: string} & T} The hashes.
+ */
```
I felt it was overkill to setup Flow-type/Typescript just for one small file, so instead I've just written the types using Typescript JSDoc syntax: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

That way most IDEs should be able to see the types/documentation, but we don't have to worry about converting this file to make valid JS.

mroz22 commented:
Hi, please rebase and feel free to include everything you need into this PR. I think it looks good so once conflicts are resolved we can merge it.

aloisklink committed:
ci(ethereum): Enable ethereumSignTypedData in CI

ethereumSignTypedData now passes on tests, so we can add it to CI.

aloisklink commented:
Thanks for the feedback! I've rebased on `develop` and `--autosquash`-ed the `!fixup` commits.

Since `develop` included #1042, the `ethereumSignTypedData` tests now pass on the new/old firmware versions that I tested, so I also added `ethereumSignTypedData` into the CI in https://github.com/trezor/connect/pull/1033/commits/5530f409b923bb9209e4858f7c0e2b25105abd79

It's ready to merge from my side :smile:

Side-note, I didn't add a CI test for [`src/js/plugins/ethereum`](https://github.com/trezor/connect/blob/2ad21b03bdfd63e54c5ca2427bfbe12cb036c1f3/src/js/plugins/ethereum/typedData.js), but I can add one in a separate PR if you think it's worth it (I'd say yes on GitHub Actions because it's free, but maybe not worth the extra CI minutes on GitLab). It would be easy to add, just a `cd src/js/plugins/ethereum && yarn install && yarn test:unit`.

mroz22 commented:
> Side-note, I didn't add a CI test for [src/js/plugins/ethereum](https://github.com/trezor/connect/blob/2ad21b03bdfd63e54c5ca2427bfbe12cb036c1f3/src/js/plugins/ethereum/typedData.js), but I can add one in a separate PR if you think it's worth it (I'd say yes on GitHub Actions because it's free, but maybe not worth the extra CI minutes on GitLab). It would be easy to add, just a cd src/js/plugins/ethereum && yarn install && yarn test:unit.

Running plugins in CI would be great. Github is a good fit for this as it runs for external contributions automatically (gitlab does not) and plugins are mostly developed by external contributors I guess. So I think we would appreciate such effort.
